src/com/rockbite/bootcamp/ShopImpl.java

```diff
@@ -0,0 +1,177 @@
+package com.rockbite.bootcamp;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+
+/**
+ * implementation of the IShop interface, represents shop API
+ * in current version it have to be a singleton to make easy to implement command pattern on it
+ */
+public class ShopImpl implements IShop {
+
+    private static ShopImpl instance;
+
+    //items taken from buyers
+    private Map<Integer, Item> itemsInCashDesk;
+
+    //products to be sold in the shop
+    private List<Product> products;
+
+    private Map<Integer, Product> soldProducts;
+
+    private ShopImpl() {
+        this.products = new ArrayList<>();
+        this.soldProducts = new HashMap<>();
+        this.itemsInCashDesk = new HashMap<>();
+    }
+
+    public static ShopImpl getInstance() {
+        if (instance == null)
+            instance = new ShopImpl();
+        return instance;
+    }
+
+    @Override
+    public List<Product> showProducts() {
+        return this.products;
+    }
+
+    /**
+     * with this method the PLayer/buyer buys a product from the shop
+     * @param buyer PLayer
+     * @param product Product which is being sold in the shop
+     * @throws IOException
+     * @throws ClassNotFoundException
+     */
+    @Override
+    public void buy(final Player buyer, final Product product) throws IOException, ClassNotFoundException {
+
+        //deep clone of Player's items
+        Map<Integer, Item> copyOfPlayerItems = new HashMap<>();
```

Comment:
this is a big overcomplication which command pattern is exactly meant to solve, but instead you are not using it an dovercomplicating purchase process.
Instead you shoul've had just normal purchase process, and refund process.

And have separate Command classes handle the history and storing as per pattern. It was not in any way meant to "modify" shop class.